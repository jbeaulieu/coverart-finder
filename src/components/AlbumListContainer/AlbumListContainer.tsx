import { Box } from "@mui/material";
import type { Album } from "../../@types/album";
import { FixedSizeList, type ListChildComponentProps } from 'react-window';
import AlbumSearchResult from "../AlbumSearchResult";

export type Props = {
  albumList: Album[];
  selectedIndex: number;
  onSelect: (selected: number) => void;
};

// Helper method that runs for each album in the albumList prop, generating a
// corresponding AlbumSearchResult component. The react-window <FixedSizeList> component
// expects a child method like this which can be run to generate the components for each
// item in the list, rather than just an array of components. This method does that while
// keeping the AlbumSearchResult component independent.
function renderRow(props: ListChildComponentProps<Props>) {
  const { index, style, data } = props;
  const { albumList, selectedIndex, onSelect } = data;
  const album = albumList[index];

  return (
    <AlbumSearchResult album={album} selected={index === selectedIndex} onSelect={onSelect} style={style} />
  );
}

// This helper function memoizes incoming props,
// To avoid causing unnecessary re-renders pure Row components.
// This is only needed since we are passing multiple props with a wrapper object.
// If we were only passing a single, stable value (e.g. albums),
// We could just pass the value directly.
// const createItemData = memoize((items, toggleItemActive) => ({
//   items,
//   toggleItemActive,
// }));

const AlbumListContainer = (props: Props) => {
  const { albumList, selectedIndex, onSelect } = props;
  return (
  <Box sx={{ width: '100%', height: 400, maxWidth: 450 }}>
    <FixedSizeList
      height={400}
      width={450}
      itemCount={albumList.length}
      itemData={{albumList, selectedIndex, onSelect}}
      itemSize={72}
      overscanCount={5}
    >
      {renderRow}
    </FixedSizeList>
  </Box>
  );
};

export default AlbumListContainer;
