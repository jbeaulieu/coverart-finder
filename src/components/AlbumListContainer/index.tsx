import AlbumListContainer from "./AlbumListContainer";

export default AlbumListContainer;
