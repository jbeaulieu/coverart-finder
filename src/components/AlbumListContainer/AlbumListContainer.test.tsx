import { fireEvent, render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import AlbumListContainer from './AlbumListContainer';
import type { Props } from './AlbumListContainer';
import type { Album } from '../../@types/album';
import { createRef} from 'react';

const album0: Album = {
  id: 1767673630,
  name: 'Watch the Fire',
  artistName: 'Dashboard Confessional',
};

const album1: Album = {
  id: 276649794,
  name: 'Bad Attitude',
  artistName: 'Meat Loaf',
}

const album2: Album = {
  id: 580265083,
  name: 'Crazy World',
  artistName: 'Boys Like Girls',
}

const onSelect = vi.fn();

const defaultProps: Props = {
  albumList: [album0, album1, album2],
  listRef: createRef(),
  selectedAlbumId: album2.id,
  onSelect,
};

const renderAlbumListContainer = (props = defaultProps) => {
  return render(<AlbumListContainer {...props} />);
};

describe('Component AlbumListContainer', () => {

  describe('with default props', () => {
    it('renders without crashing', () => {
      const { asFragment } = renderAlbumListContainer();

      expect(asFragment()).toMatchSnapshot();
    });

    it('applies the "selected" prop correctly', () => {
      renderAlbumListContainer();

      const nonSelectedItem1 = screen.getByRole("button", { name: `${album0.name} ${album0.artistName}` });
      const nonSelectedItem2 = screen.getByRole("button", { name: `${album1.name} ${album1.artistName}` });
      const selectedItem = screen.getByRole("button", { name: `${album2.name} ${album2.artistName}` });

      expect(nonSelectedItem1).not.toHaveClass("Mui-selected");
      expect(nonSelectedItem2).not.toHaveClass("Mui-selected");
      expect(selectedItem).toHaveClass("Mui-selected");
    });
  });

  describe('on click', () => {
    it('invokes onSelect function', () => {
      renderAlbumListContainer();

      const itemButton = screen.getByRole("button", { name: `${album1.name} ${album1.artistName}` });

      fireEvent.click(itemButton);

      expect(onSelect).toHaveBeenCalledExactlyOnceWith(album1.id);
    });
  });
});
