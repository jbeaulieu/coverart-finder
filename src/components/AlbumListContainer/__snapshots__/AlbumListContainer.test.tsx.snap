// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component AlbumListContainer > on click > renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1l9xq7o"
  >
    <div
      style="position: relative; height: 400px; width: 450px; overflow: auto; -webkit-overflow-scrolling: touch; will-change: transform; direction: ltr;"
    >
      <div
        style="height: 216px; width: 100%;"
      >
        <li
          class="MuiListItem-root MuiListItem-gutters css-19f6boz-MuiListItem-root"
          style="position: absolute; left: 0px; top: 0px; height: 72px; width: 100%;"
        >
          <div
            class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline css-14ln1j6-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
              >
                Watch the Fire
              </span>
              <p
                class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-1a1whku-MuiTypography-root"
              >
                Dashboard Confessional
              </p>
            </div>
            <div
              class="MuiListItemAvatar-root css-1ulznza-MuiListItemAvatar-root"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-18k2bs-MuiAvatar-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-1mo2pzk-MuiSvgIcon-root-MuiAvatar-fallback"
                  data-testid="PersonIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </li>
        <li
          class="MuiListItem-root MuiListItem-gutters css-19f6boz-MuiListItem-root"
          style="position: absolute; left: 0px; top: 72px; height: 72px; width: 100%;"
        >
          <div
            class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline css-14ln1j6-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
              >
                Bad Attitude
              </span>
              <p
                class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-1a1whku-MuiTypography-root"
              >
                Meat Loaf
              </p>
            </div>
            <div
              class="MuiListItemAvatar-root css-1ulznza-MuiListItemAvatar-root"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-18k2bs-MuiAvatar-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-1mo2pzk-MuiSvgIcon-root-MuiAvatar-fallback"
                  data-testid="PersonIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </li>
        <li
          class="MuiListItem-root MuiListItem-gutters css-19f6boz-MuiListItem-root"
          style="position: absolute; left: 0px; top: 144px; height: 72px; width: 100%;"
        >
          <div
            class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters Mui-selected MuiListItemButton-root MuiListItemButton-gutters Mui-selected css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline css-14ln1j6-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
              >
                Crazy World
              </span>
              <p
                class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-1a1whku-MuiTypography-root"
              >
                Boys Like Girls
              </p>
            </div>
            <div
              class="MuiListItemAvatar-root css-1ulznza-MuiListItemAvatar-root"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-18k2bs-MuiAvatar-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-1mo2pzk-MuiSvgIcon-root-MuiAvatar-fallback"
                  data-testid="PersonIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </li>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Component SearchContainer > on click > renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1l9xq7o"
  >
    <div
      style="position: relative; height: 400px; width: 450px; overflow: auto; -webkit-overflow-scrolling: touch; will-change: transform; direction: ltr;"
    >
      <div
        style="height: 216px; width: 100%;"
      >
        <li
          class="MuiListItem-root MuiListItem-gutters css-19f6boz-MuiListItem-root"
          style="position: absolute; left: 0px; top: 0px; height: 72px; width: 100%;"
        >
          <div
            class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline css-14ln1j6-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
              >
                Watch the Fire
              </span>
              <p
                class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-1a1whku-MuiTypography-root"
              >
                Boys Like Girls
              </p>
            </div>
            <div
              class="MuiListItemAvatar-root css-1ulznza-MuiListItemAvatar-root"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-18k2bs-MuiAvatar-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-1mo2pzk-MuiSvgIcon-root-MuiAvatar-fallback"
                  data-testid="PersonIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </li>
        <li
          class="MuiListItem-root MuiListItem-gutters css-19f6boz-MuiListItem-root"
          style="position: absolute; left: 0px; top: 72px; height: 72px; width: 100%;"
        >
          <div
            class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline css-14ln1j6-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
              >
                Boys Like Girls
              </span>
              <p
                class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-1a1whku-MuiTypography-root"
              >
                Boys Like Girls
              </p>
            </div>
            <div
              class="MuiListItemAvatar-root css-1ulznza-MuiListItemAvatar-root"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-18k2bs-MuiAvatar-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-1mo2pzk-MuiSvgIcon-root-MuiAvatar-fallback"
                  data-testid="PersonIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </li>
        <li
          class="MuiListItem-root MuiListItem-gutters css-19f6boz-MuiListItem-root"
          style="position: absolute; left: 0px; top: 144px; height: 72px; width: 100%;"
        >
          <div
            class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters Mui-selected MuiListItemButton-root MuiListItemButton-gutters Mui-selected css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline css-14ln1j6-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
              >
                Crazy World
              </span>
              <p
                class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-1a1whku-MuiTypography-root"
              >
                Boys Like Girls
              </p>
            </div>
            <div
              class="MuiListItemAvatar-root css-1ulznza-MuiListItemAvatar-root"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-18k2bs-MuiAvatar-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-1mo2pzk-MuiSvgIcon-root-MuiAvatar-fallback"
                  data-testid="PersonIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </li>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
